extstore: support direct I/O and async I/O

mita opened the pull request:
This adds support for direct I/O and async I/O on extstore.

With `-o ext_direct` at startup, memcached opens the extstore files with
the O_DIRECT flag.  When direct I/O is used, extstore reads and writes
bypass the operating system caches.

To build with async I/O support you will have to install io_uring library
(http://git.kernel.dk/cgit/liburing/) and ./configure --enable-liburing.
With `-o ext_io_engine=io_uring` and I/O depth > 1
(e.g. `-o ext_io_depth=64), the IO threads execute each IO object on their
queues in asynchronous manner.

This maximizes I/O throughput with smaller number of the IO threads when
direct I/O is used.

dormando commented:
Thanks! This is fun, I've been wanting to do both of these things for a while but my priorities are elsewhere.

Might take a while to review due to time right now. Do you have any performance test results for the various modes, etc? Not a lot of comments in the patch and a lot of new things are going on in there.

With io_uring I was hoping to replace the entire event loop so extstore requests don't have to push to a separate thread anymore, but of course that's a lot more work.

mita commented:
In order to make it easy to review, I plan to split this single change into at least three smaller changes.

* Add extstore I/O engine infrastructure
* Support direct I/O
* Support async I/O


I usually do performance tests against extstore intensive read-only workloads.

```
# start memcached
memcached -t 16 -c 48000 -m 40000 -b 4096 \
  -o ext_wbuf_size=64,ext_item_size=240,ext_max_frag=0.80,hashpower=30,no_lru_crawler,ext_recache_rate=0,ext_item_age=5,ext_path=/mnt/ssd/extstore:200G
# mget benchmark (1000 mget, 8 io threads, 256 bytes value, 10000000 requests)
memstrike -d 127.0.0.1:11211 -m 1000 -t 8 -v 256 10000000
```

The ext_path is a file in ext4 file system on Intel SSD 900P.

With additional extended options `-o ext_threads=4,ext_io_engine=io_uring,ext_direct,ext_io_depth=64`
(direct I/O, async I/O, and four IO threads): 527712.951173 req/sec
This saturates the SSD bandwidth.

With additional extended options`-o ext_threads=20,ext_io_engine=sync,ext_direct`
(direct I/O, normal blocking I/O, and twenty IO threads): 483425.397672 req/sec

pkarumanchi9 reviewed (changes requested):
thanks for doing this change-set. I've done one pass and left some comments for review. 
The usage information is missing in memcached.c file. Can you add that as well?

mita commented on extstore.c:
```diff
@@ -360,6 +455,13 @@ void *extstore_init(struct extstore_conf_file *fh, struct extstore_conf *cf,
+        if (e->ops->thread_init) {
+            *res = e->ops->thread_init(&e->io_threads[i]);
+            if (*res) {
+                free(e);
```
Fixed by commit 884e3b11028c6a8756ad4bc921472d5c0beaeb2b
The change also releases e->wbuf_stack and e->io_stack.

mita commented on extstore.c:
```diff
@@ -350,7 +421,31 @@ void *extstore_init(struct extstore_conf_file *fh, struct extstore_conf *cf,
+
+    if ((e->max_io_size % e->io_align) != 0 || !mc_powerof2(e->io_align)) {
+        *res = EXTSTORE_INIT_BAD_IO_SIZE;
+        free(e);
```
Fixed by commit 884e3b11028c6a8756ad4bc921472d5c0beaeb2b

mita commented on extstore.c:
```diff
@@ -710,6 +812,494 @@ static inline int _read_from_wbuf(store_page *p, obj_io *io) {
+
+    ret = io_uring_queue_init(e->io_depth, &priv->ring, 0);
+    if (ret < 0) {
+        E_DEBUG("io_uring_queue_init: %s\n", strerror(-ret));
```
Fixed by commit 884e3b11028c6a8756ad4bc921472d5c0beaeb2b

mita commented on extstore.c:
```diff
@@ -710,6 +812,494 @@ static inline int _read_from_wbuf(store_page *p, obj_io *io) {
+        struct store_io_data *data = malloc(sizeof(*data));
+
+        if (e->direct) {
+            if (posix_memalign(&data->read_buf, e->io_align, e->max_io_size)) {
```
Fixed by commit 884e3b11028c6a8756ad4bc921472d5c0beaeb2b

mita commented on extstore.c:
```diff
@@ -710,6 +812,494 @@ static inline int _read_from_wbuf(store_page *p, obj_io *io) {
+
+    extstore_io_done(t, io, ret);
+
+    return false;
```
Updated comment about inflight in struct store_io_thread:

unsigned int inflight; // number of inflight I/O requests, this is used only for async I/O

mita commented on extstore.c:
```diff
@@ -710,6 +812,494 @@ static inline int _read_from_wbuf(store_page *p, obj_io *io) {
+
+static void extstore_io_submit(store_io_thread *t, obj_io *io) {
+    if (t->e->ops->submit(t, io))
+        t->inflight++;
```
t->inflight is only updated by the thread `t', so it doesn't need to be protected by mutex.

mita commented on extstore.c:
```diff
@@ -710,6 +812,494 @@ static inline int _read_from_wbuf(store_page *p, obj_io *io) {
+static void extstore_io_complete(store_io_thread *t) {
+    if (t->e->ops->complete) {
+        if (t->e->ops->complete(t))
+            t->inflight--;
```
Same as above.

mita commented on extstore.c:
```diff
@@ -73,15 +87,56 @@ typedef struct _store_page {
+    uint64_t iov_offset;
+    int fd;
+    void *read_buf;
+    struct iovec iov_oneshot;
```
I've added comment to clarify:
/* this is used when io->iov is NULL or resubmitting after short read/write */

tharanga commented:
@mita we are excited about this change. What are the liburing and kernel versions you used?

mita commented:
> @mita we are excited about this change. What are the liburing and kernel versions you used?

I get the latest liburing (0.7+) and build from source on Ubuntu 18.04 LTS. 
The kernel version is 5.4.0-42-generic.

tharanga commented:
> > @mita we are excited about this change. What are the liburing and kernel versions you used?
> 
> I get the latest liburing (0.7+) and build from source on Ubuntu 18.04 LTS.
> The kernel version is 5.4.0-42-generic.

Thanks! I saw the liburing is updated to match with the kernel 5.7 and wondering it's compatibility.

tharanga commented on configure.ac:
```diff
@@ -228,6 +234,7 @@ AM_CONDITIONAL([ENABLE_SASL],[test "$enable_sasl" = "yes"])
 AM_CONDITIONAL([ENABLE_EXTSTORE],[test "$enable_extstore" != "no"])
 AM_CONDITIONAL([ENABLE_ARM_CRC32],[test "$enable_arm_crc32" = "yes"])
 AM_CONDITIONAL([ENABLE_TLS],[test "$enable_tls" = "yes"])
+AM_CONDITIONAL([ENABLE_LIBURING],[test "$enable_liburing" = "yes"])
```
If you haven't done already, it's a good idea to run extstore tests in the async mode. You may do something similar to TLS (https://github.com/memcached/memcached/blob/master/Makefile.am#L128), run all extstore tests with io_uring.

mita replied:
Sounds good.

Commit 5dee70a0bb65ec921307aab5d3f58b4673b10f98 introduced a new environment variable which can be used to test with additional memcached server options.

mita commented on extstore.c:
```diff
@@ -710,6 +812,494 @@ static inline int _read_from_wbuf(store_page *p, obj_io *io) {
+    data = io_uring_cqe_get_data(cqe);
+
+    ret = cqe->res;
+    if (ret <= 0) {
```
Presumably 'ret == 0' can only happen when reading the end of file (e.g. the extstore is accidentally truncated).
So I would like to avoid infinite loop by this condition.

mita commented on extstore.c:
```diff
@@ -710,6 +812,494 @@ static inline int _read_from_wbuf(store_page *p, obj_io *io) {
+    if (done) {
+        ret = data->io->len;
+    } else {
+        struct io_uring_sqe *sqe = io_uring_get_sqe(&priv->ring);
```
Fixed by commit f2328d482457acb53fa89f3d241fc05859437bbf

dormando commented:
What's memstrike from?

I'm still a few days behind being able to look closely. I like the idea of splitting up the patch; I'm very likely to take the direct io patch. I probably want to at least keep the io_uring one around for reference but I had different plans for it. A closer look will help me figure this out though. I feel like there are also too many tunables now.

For direct io at least, you'll probably see the biggest differences in performance under mixed read/write workloads. The performance could go either way.

mita commented:
> What's memstrike from?

The memstrike is a memcached benchmark tool which supports multiple server and mget, but does not support read/write mixed workloads.

https://github.com/frsyuki/memstrike

I have split up the direct IO and async IO patch into three patches.

      extstore: support direct I/O
      extstore: introduce extstore IO engine infrastructure
      extstore: support async I/O

dormando commented:
Hey,

I fell off the earth for a while. This is a pretty involved change so I'm not going to get away with some quick testing :) 

What would be nice if you or anyone else wants to, please do mixed read/write workload tests. Use mc-crusher(https://github.com/memcached/mc-crusher) and create some test configs to share here. You can use the test-suites code to design a quick suite of tests which will do ramp-to-failure and other interesting things. I've had to do all this (and more!) while working on larger changes like this.

The tunables I'd have to think about one by one; I'm trying pretty hard to not add more of those and keep sensible defaults. It's not quite obvious to me from a skim what those might be.

As far as plans go, I'm hoping to have an entire io_uring event loop. Then extstore requests can get executed internally to each worker and we avoid the sidethread push and multiple syscalls on pipes. This code still helps a lot as a point of reference at very least, and I hope I can take the direct io code still.

mita commented:
I've created a test script based on test-suite-example.

https://github.com/mita/mc-crusher/blob/direct_io/test-suites/test-direct

In addition to the normal extstore test and RAM test without extstore,This script adds direct IO and direct IO + io_uring test configurations.

I agree that extstore requests with io_uring should not be pushed to a separate thread.  I plan to work on this.

mita committed:
extstore: introduce extstore IO engine infrastructure

mita committed:
Add ability to test with additional memcached options

This adds a new MEMCACHED_EXTRA_ARGS environment variable for running
tests with additional memcached server options.

A test with direct I/O:

MEMCACHED_EXTRA_ARGS="-o ext_direct" make test

A test with direct I/O and async I/O:

MEMCACHED_EXTRA_ARGS="-o ext_io_engine=io_uring,ext_direct,ext_io_depth=64" make test

mita committed:
extstore: support direct I/O

This adds support for direct I/O on extstore.

With `-o ext_direct` at startup, memcached opens the extstore files with
the O_DIRECT flag.  When direct I/O is used, extstore reads and writes
bypass the operating system caches.

mita committed:
extstore: support async I/O

This adds support for async I/O on extstore.

To build with async I/O support you will have to install io_uring library
(http://git.kernel.dk/cgit/liburing/) and ./configure --enable-liburing.
With `-o ext_io_engine=io_uring` and I/O depth > 1
(e.g. `-o ext_io_depth=64), the worker threads directly read and write
asynchronouly without the IO threads.

mita commented:
The latest version does not use the I/O threads when the io_uring engine is selected.
The worker threads directly submit I/O requests to extstore asynchronously.

mita commented:
This table shows performance comparison between default sync engine and io_uring engine.
The io_uring engine gets better throughput (requets/sec) with smaller number of client threads.

|client threads|2|4|8|12|16|
|--------------|-|-|-|--|--|
|sync engine                |338768|712797|1145488|1351642|1323820|
|io_uring engine            |455830|855805|1266225|1384047|1323683|
|sync engine, direct I/O    |95181|177153|305078|419204|490403|
|io_uring engine, direct I/O|289145|481540|507430|511633|512510|

Benchmark details:

1. Run memcached

```
./memcached/memcached -t 16 -m 8000 -o ext_item_age=0,ext_path=/mnt/ssd/extstore:200G $EXTRA_ARGS
```

The value of EXTRA_ARGS varies depending on the configuration.

* sync engine: `-o ext_io_engine=sync,ext_threads=16,ext_io_depth=64`
* io_uring engine: `-o ext_io_engine=io_uring,ext_io_depth=64`
* sync engine, direct I/O: `-o ext_io_engine=sync,ext_direct,ext_threads=16,ext_io_depth=64`
* io_uring engine, direct I/O: `-o ext_io_engine=io_uring,ext_direct,ext_io_depth=64`

The ext_path is a file in ext4 file system on Intel SSD 900P.

2. Load 10,000,000 records.

```
./memstrike/memstrike -d '127.0.0.1:11211' -t 16 -v 512 10000000 -s
```

3. Wait for extstore settle

```
./mc-crusher/bench-sample 127.0.0.1:11211 0 2 extstore_objects_written
```

4. Run read only workloads with different number of client threads

```
for i in 2 4 8 12 16; do
  ./memstrike/memstrike -d '127.0.0.1:11211' -m 10 -t $i -v 512 10000000 -g
done
```

dormando commented:
Whoops I missed this update. Any idea why direct IO is so much slower?

mita commented:
> Whoops I missed this update. Any idea why direct IO is so much slower?

Thanks for looking at update.

The direct IO performance is limited by the random read IOPS.
These results are measured with a single Intel SSD 900P.
We'll get better numbers with using multiple devices for extstore.

dormando commented:
@mita did you end up using this in a production context anywhere?

I've been both too busy and too-allocated to do the work necessary to validate this change, but that might be changing. I finally spent some time with io_uring and have more thoughts on it now as well.

lidaohang commented:
mark
